Better errors when a timeout occurs.

AaronPatterson opened the pull request:
## Description
When setting a timeout on the HttpClient used by the SDK’s and a timeout occurs a TaskCanceledException is surfaced to the consumer.  Frequently when these types of timeouts occur, it is due to one of the retry handlers (RetryAfterDelegatingHandler or RetryDelegatingHandler) doing their job and retrying requests which are failing.  Ideally information about the failure which was occurring would be returned to the consumer as this is more meaningful/actionable than a TaskCanceledException.

This set of changes updates RetryAfterDelegatingHandler and RetryDelegatingHandler. To ensure that they return the last error they received when cancellation happens.  Doing this ensures that the consumer gets back a more meaningful response when the timeout is hit.

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] Please add REST spec PR link to the SDK PR
- [X ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ X] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [X ] Title of the pull request is clear and informative.
- [ X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [ X] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

AaronPatterson committed:
Better Error when a timeout occurs.

When setting a timeout on the HttpClient used by the SDK’s and a timeout occurs a TaskCanceledException is surfaced to the consumer.  Frequently when these types of timeouts occur, it is due to one of the retry handlers (RetryAfterDelegatingHandler or RetryDelegatingHandler) doing their job and retrying requests which are failing.  Ideally information about the failure which was occurring would be returned to the consumer as this is more meaningful/actionable than a TaskCanceledException.

This set of changes updates RetryAfterDelegatingHandler and RetryDelegatingHandler. To ensure that they return the last error they received when cancellation happens.  Doing this ensures that the consumer gets back a more meaningful response when the timeout is hit.

shahabhijeet reviewed (changes requested):
I see you have deleted existing tests.
You have undone a perf bug that was fixed for memory leak due to multiple subscription of retry policy events. (I do not see any explanation for undoing it)

At this point this seems like a big change and this affects not only .NET SDKs but also powershell as well as Fluent SDKs

I will need more information and more time to go through this propose change.
Especially a breaking change prior to //BUILD will be very hard to pull off for internal dependencies as well as for customers.

AaronPatterson commented on src/SdkCommon/ClientRuntime/Tests/ClientRuntime.NetCore.Tests/TransientFaultHandling/MemoryTest/CRMemoryTests.cs:
```diff
@@ -1,31 +0,0 @@
-    public class CRMemoryTests
-    {
-        [Fact]
-        public void MeasureSvcClientMem()
```
The tests were testing that subscriptions were removed.  Now subscriptions are never made so there is nothing left to test.

AaronPatterson commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryDelegatingHandler.cs:
```diff
@@ -39,7 +35,7 @@ public RetryDelegatingHandler() : base()
         /// </summary>
         /// <param name="innerHandler">Inner http handler.</param>
-        public RetryDelegatingHandler(DelegatingHandler innerHandler)
+        public RetryDelegatingHandler(HttpMessageHandler innerHandler)
```
This is not a breaking change.  DelegatingHandler derives from HttpMessageHandler, so this makes the these types more flexible.

The reason I changed this is so I could use the "FakeHttpHandler" in testing these handlers.

fearthecowboy replied:
This is a binary breaking change. You can not drop in a new compiled binary with existing binaries with this change. 
If you recompiled the code that used the library, yes, this would compile and function, but it is most definitely breaking change, and cannot be transparently rolled out on top of existing binaries.
​
Attempting to do so will produce:
`Unhandled Exception: System.MissingMethodException: Method not found: 'Void RetryDelegatingHandler..ctor(System.Net.Http.DelegatingHandler)'. `

AaronPatterson replied:
Changed this to use a new constructor overload to avoid the binary breaking change.

AaronPatterson committed:
Avoid breaking binary compatability by adding new constructor overload instead of changing to base type directly.

shahabhijeet reviewed (changes requested):
@AaronPatterson can you resolve merge conflicts first before I can review.

AaronPatterson committed:
Merge remote-tracking branch 'origin/master' into ImproveTimeoutErrors

shahabhijeet commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryDelegatingHandler.cs:
```diff
@@ -117,15 +99,27 @@ public int EventCallbackCount
                     if (!responseMessage.IsSuccessStatusCode)
                     {
-                        // dispose the message unless we have stopped retrying
-                        this.Retrying += (sender, args) =>
```
@AaronPatterson I see existing code has been deleted.
This PR cannot be accepted when existing functionality/code is being deleted.
Please add the feature without deleting any code or removing any existing functionality.

AaronPatterson replied:
What was being done previously was that a delegate was being registered each time a request was failed and before a retry happens the delegates would get invoked to allow them to do any cleanup they wanted to do.  The reason this was changed is that it was disposing of the previous response before it got a new one and this was one of the things which was leading to TaskCanceledExceptions instead of getting back the previously received error.  The updated code still does the disposing of the previous response, but it waits to do this until it has a new response.  With the new approach, the private Retrying delegate is no longer needed.

All of the externally observable functionality is still in place, so I do not believe that this is removing any functionality.

shahabhijeet commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryDelegatingHandler.cs:
```diff
@@ -139,51 +133,15 @@ public int EventCallbackCount
-        /// <summary>
-        /// An instance of a callback delegate that will be invoked whenever a retry condition is encountered.
-        /// </summary>
-        private event EventHandler<RetryingEventArgs> Retrying;
```
You cannot remove the event that is part of our existing extensibility provided via RetryPolicy.
This event is wired up via RetryPolicy that customers can set according to their needs.
Please have all existing functionality back and then have the required code change.

AaronPatterson replied:
The EventHandler being removed here is private and does not impact the existing extensibility provided by the RetryPolicy.

shahabhijeet commented on src/SdkCommon/ClientRuntime/Tests/ClientRuntime.NetCore.Tests/TransientFaultHandling/MemoryTest/CRMemoryTests.cs:
```diff
@@ -1,31 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License. See License.txt in the project root for license information.
-
-namespace CR.Azure.NetCore.Tests.MemoryTest
```
Please do not delete existing test.

AaronPatterson replied:
The test was previously testing that handlers are not being leaked by the RetryDelegatingHandler.  Since the RetryDelegatingHandler is not registering handlers anymore the test seemed obsolete, thus the removal.

I will bring back the test and update it to test the new behavior.  Now that the RetryDelegatingHandler is not making use of the Retrying handler, the test will validate that no handlers are being registered.

shahabhijeet reviewed (changes requested):
1. I see ClientRuntime tests are failing.
2. You can see which tests are failing by opening the .trx file from the CI job link under Build artifacts.
3. Please do not remove existing functionality (details have been added in the PR review).
4. Please add back the deleted test.

AaronPatterson committed:
Reintroduce the EventCallbackCount property on the RetryDelegatingHandler to avoid breaking binary compatibility.

The property originally returned the count of handlers registered on the private Retrying EventHandler.  Since there is no longer a need for the private Retrying EventHandler, updating to return the count of handlers on the RetryPolicy since it is the closest meaningful value.

AaronPatterson committed:
Updating RetryAfterDelegatingHandlerTests to have a timeout in seconds now that the handler is treating the "Retry-After" in seconds.

AaronPatterson commented:
@shahabhijeet I pushed updates to the tests and one additional update to avoid breaking binary compatibility.

AaronPatterson committed:
Bring back the CRMemoryTests

The test was previously testing that handlers are not being leaked by the RetryDelegatingHandler.

Now that the RetryDelegatingHandler is not making use of the Retrying handler, the test validates that no handlers are being registered.

markcowl commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryDelegatingHandler.cs:
```diff
@@ -139,51 +138,15 @@ public int EventCallbackCount
                 return responseMessage;
             }
-            catch(Exception ex)
+            catch (Exception)
```
use a when clause here so we don't rethrow the error unless necessary

markcowl commented on src/SdkCommon/ClientRuntime/Tests/TestCommon/Fakes/FakeHttpHandler.cs:
```diff
@@ -27,10 +28,18 @@ public FakeHttpHandler()
-            var response = new HttpResponseMessage(HttpStatusCode.OK);
+            var response = new HttpResponseMessage(HttpStatusCode.OK)
+            {
+                Content = new StringContent(String.Empty)
```
We should make sure we still have tests with null Content : the above code handles it, but this could regress in the future.

AaronPatterson committed:
Update tests to test for null content in the response.

Avoid throwing by using when clause RetryDelegatingHandler.

shahabhijeet commented:
@AlexGhiondea can you approve this PR as well, from KeyVault data plane point of view, let me know if you see any concerns.

AlexGhiondea commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryAfterDelegatingHandler.cs:
```diff
@@ -44,35 +53,72 @@ public RetryAfterDelegatingHandler(DelegatingHandler innerHandler)
+                    response = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+                    // if they send back a 429
+                    if (response.StatusCode == (HttpStatusCode)429)
```
Nit: these 2 if statements could be collapsed to reduce indentation.

AaronPatterson replied:
Was trying to leave the existing code alone as much as possible, but I'm not opposed to cleaning it up some.

AlexGhiondea replied:
Ah -- didn't realize this was existing code -- I am fine with not churning that code in this PR.

AlexGhiondea commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryAfterDelegatingHandler.cs:
```diff
@@ -44,35 +53,72 @@ public RetryAfterDelegatingHandler(DelegatingHandler innerHandler)
                         }
                     }
                 }
+                catch (TaskCanceledException) when(previousResponseMessage != null)
```
nit: space between `when` and `(`

AlexGhiondea commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryAfterDelegatingHandler.cs:
```diff
@@ -44,35 +53,72 @@ public RetryAfterDelegatingHandler(DelegatingHandler innerHandler)
-                            // wait for that duration
-                            await Task.Delay(TimeSpan.FromSeconds(retryAfter), cancellationToken);
+                        // and there is a retry-after header
+                        if (response.Headers.Contains("Retry-After"))
```
Why not capture the content for all the attempts? Could there be useful information in the previous responses that we are discarding?

AaronPatterson replied:
In the end we can only return 1 response and the latest response has the most up to date information.

AlexGhiondea replied:
I was wondering if we could concatenate all the responses and set the concatenated response as the one we send back -- but I am fine with just using the last one

AlexGhiondea commented on src/SdkCommon/ClientRuntime/ClientRuntime/RetryAfterDelegatingHandler.cs:
```diff
@@ -44,35 +53,72 @@ public RetryAfterDelegatingHandler(DelegatingHandler innerHandler)
                 // if we haven't hit continue, then return the response up the stream
-                return response;
-            } while( true );
+                return response ?? previousResponseMessage;
```
This can still return null, right?

AaronPatterson replied:
We will always have a response or previousResponseMessage here.  It is possible that cancellation occurs before we get any response and in that case the TaskCanceledException will be allowed to bubble up and we will never get here.

AlexGhiondea replied:
What happens if the error happened in the `SendAsync` method and that throws something other than `TaskCanceledException`?

AaronPatterson replied:
That exception will bubble up and this line will never be executed.

AlexGhiondea reviewed (changes requested):
I left some comments about this change.

AaronPatterson committed:
- Cleanup in RetryAfterDelegatingHandler
- Remove changes from FakeHttpHandler that were not not needed.

AaronPatterson committed:
Merge remote-tracking branch 'origin/master' into ImproveTimeoutErrors

AaronPatterson commented:
I've made all of the requested updates.
